lib/pg_ha_migrations/safe_statements.rb

```diff
@@ -170,6 +170,65 @@ def safe_remove_concurrent_index(table, options={})
     unsafe_remove_index(table, **options.merge(:algorithm => :concurrently))
   end
 
+  def safe_add_concurrent_partitioned_index(table, columns, name_suffix: nil, if_not_exists: false)
+    if ActiveRecord::Base.connection.postgresql_version < 11_00_00
+      raise PgHaMigrations::InvalidMigrationError, "Concurrent partitioned index creation not supported on Postgres databases before version 11"
+    end
+
+    schema, table = _schema_and_table_for(table)
+
+    is_partitioned = select_value(<<~SQL)
+      SELECT EXISTS (
+        SELECT 1
+        FROM pg_partitioned_table, pg_class, pg_namespace
+        WHERE pg_class.oid = pg_partitioned_table.partrelid
+          AND pg_class.relnamespace = pg_namespace.oid
+          AND pg_class.relname = #{quote(table)}
+          AND pg_namespace.nspname = #{quote(schema)}
+      )
+    SQL
+
+    raise PgHaMigrations::InvalidMigrationError, "Table #{table} is not a partitioned table" unless is_partitioned
+
+    name_suffix = "#{Array.wrap(columns).join("_")}_idx" unless name_suffix.present?
+
+    child_tables = select_values(<<~SQL)
+      SELECT child.relname
+      FROM pg_inherits
+        JOIN pg_class parent ON pg_inherits.inhparent = parent.oid
+        JOIN pg_class child  ON pg_inherits.inhrelid  = child.oid
+        JOIN pg_namespace    ON parent.relnamespace = pg_namespace.oid
+      WHERE parent.relname = #{quote(table)}
+        AND pg_namespace.nspname = #{quote(schema)}
+    SQL
+
+    parent_index = "#{table}_#{name_suffix}"
```

Comment:
Do we need to handle long names like ActiveRecord does?

Reply:
Oh I didn't realize there was a character limit, so probably, yes. I'll try to look at the rails code to see what they're doing.